tools/rosgraph/src/rosgraph/roslogging.py

```diff
@@ -39,12 +39,31 @@
 import time
 import logging
 import logging.config
+import inspect
 
 import rospkg
 from rospkg.environment import ROS_LOG_DIR
 
 class LoggingException(Exception): pass
 
+class RospyLogger(logging.Logger):
+    def findCaller(self):
+        """
+        Find the stack frame of the caller so that we can note the source
+        file name, line number, and function name with class name if possible.
+        """
+        frame, _, lineno, _, _, _ = inspect.stack()[3]
+        file_name = inspect.getabsfile(frame)
+        func_name = frame.f_code.co_name
+        try:
+            class_name = frame.f_locals['self'].__class__.__name__
+            func_name = '%s.%s' % (class_name, func_name)
+        except KeyError:
+            pass
+        return file_name, lineno, func_name
+
+logging.setLoggerClass(RospyLogger)
```

Comment:
This should ensure not to overwrite custom logger classes already being set by user code. See https://docs.python.org/2/library/logging.html#logging.getLoggerClass on how to achieve this.